snuba/query/data_source/join.py

```diff
@@ -0,0 +1,114 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from enum import Enum
+from typing import Generic, Mapping, NamedTuple, Optional, Sequence, Union
+
+from snuba.clickhouse.columns import ColumnSet, QualifiedColumnSet, SchemaModifiers
+from snuba.query import ProcessableQuery, TSimpleDataSource
+from snuba.query.data_source import DataSource
+
+
+class JoinType(Enum):
+    INNER = "INNER"
+    LEFT = "LEFT"
+
+
+class JoinModifier(Enum):
+    ANY = "ANY"
+    SEMI = "SEMI"
+
+
+@dataclass(frozen=True)
+class JoinNode(ABC, Generic[TSimpleDataSource]):
+    """
+    Assigns an alias to a node in a join expression.
+    The alias is used in the join condition and in all the expressions
+    that rely on the join in the query.
+
+    Join nodes can be simple data sources, join expressions and queries.
+    """
+
+    alias: Optional[str]
```

Comment:
I was thinking that maybe this alias should be auto generated? There isn't really a need to pass this alias in. It could also help avoid bugs with overlapping alias names.

Reply:
Don't we allow the user to provide their own alias in the language for joins ?
```
MATCH (e: Events) -[.....]-> (g: Groups)
```

These aliases are `e` and `g`

If we override what the user provided with something internal I would be ok, but that should be done by a dedicated pass on the query after the parsing and it cannot be done unilaterally by the data source since expressions in the query need to reference these aliases and the data source does not have access to expressions in the query.